Feature/reflect typings

NoNameProvided opened the pull request:
This PR add typings for reflect-metadata and fixes a failing test because if type mismatch.

Tests are passing, no behavioral changes were made.

NoNameProvided commented:
Ok, so 0.11 and 0.12 fails, but the error is not related to the lib, but one of the dependencies using the `const` keyword which errors out.

Shouldn't we remove the 0.11 and 0.12? Does anyone out-there using this with those versions?

pleerock commented:
I think we can drop all old node versions.

pleerock commented:
please remove package-lock.json

pleerock commented on package.json:
```diff
@@ -32,6 +32,7 @@
     "@types/mocha": "^2.2.33",
     "@types/node": "0.0.2",
     "@types/sinon": "^2.2.2",
+    "@types/reflect-metadata": "0.0.5",
```
why do we need this while original reflect-metadata package provides typings in its own package?

pleerock commented on src/decorators.ts:
```diff
@@ -22,7 +22,7 @@ export function Transform(transformFn: (value: any) => any, options?: TransformO
 export function Type(typeFunction?: (type?: TypeOptions) => Function) {
     return function(target: any, key: string) {
-        const type = (Reflect as any).getMetadata("design:type", target, key);
+        const type = Reflect.getMetadata("design:type", target, key);
```
this will cause problems on users because they don't have `Reflect` with such methods until they install specific packages. But I don't them to have compiler errors from this package right after they install it.

NoNameProvided commented:
> please remove package-lock.json

Is there a way you are not adding it? Also what is the reason not pinning down the dependencies? This way every install will be slightly different.

> I think we can drop all old node versions.

I have removed 0.11 and 0.12 from travis config.

> why do we need this while original reflect-metadata package provides typings in its own package?

I have removed it. VS Code complained about missing types and I saw it's casted to any, so I have thought there were no typings for it. 

> this will cause problems on users because they don't have Reflect with such methods until they install specific packages. 

Why don't we  add it as a peer dependency then?

cojack commented:
@NoNameProvided I guess I can close this PR, as outdated.
